packages/core/nano/src/core.ts

```diff
@@ -71,47 +79,57 @@ export function makeTracer(): [Pending<unknown>[], Trace] {
     return [data, fn];
 }
 
-/** Lifting of core operations into the `Delayed` S.A.F. */
-export interface LiftedCore {
-    read: <O>(trace: Trace, host: O, context: Delayed<CalcValue<O>>, prop: string) => Delayed<CalcValue<O>>;
-    select: <L, R>(cond: Delayed<boolean>, l: () => Delayed<L>, r: () => Delayed<R>) => Delayed<L | R>;
-    app1: <O, A, B>(trace: Trace, host: O, op: <O>(trace: Trace, host: O, expr: A) => B, expr: Delayed<A>) => Delayed<B>;
-    app2: <O, A, B, C>(trace: Trace, host: O, op: <O>(trace: Trace, host: O, l: A, r: B) => C, l: Delayed<A>, r: Delayed<B>) => Delayed<C>;
-    appN: <O>(trace: Trace, host: O, fn: Delayed<CalcValue<O>>, args: Delayed<CalcValue<O>>[]) => Delayed<CalcValue<O>>;
+/**
+ * Core expression runtime that implements collection and propagation
+ * of potentially unavailable resources.
+ *
+ * This is like a selective applicative functor + fetch, if you squint
+ * hard enough. `app1` and `app2` are specialised to the cases where
+ * the function is always produced via `pure`.
+ */
+export interface Runtime {
+    read: <O>(origin: O, context: Delayed<CalcValue<O>>, prop: string) => Delayed<CalcValue<O>>;
+    ifS: <A>(cond: Delayed<boolean>, cont: (cond: boolean) => Delayed<A>) => Delayed<A>;
+    app1: <O, A, B>(origin: O, op: (runtime: Runtime, origin: O, expr: A) => B, expr: Delayed<A>) => Delayed<B>;
+    app2: <O, A, B, C>(origin: O, op: (runtime: Runtime, origin: O, l: A, r: B) => C, l: Delayed<A>, r: Delayed<B>) => Delayed<C>;
+    appN: <O>(origin: O, fn: Delayed<CalcValue<O>>, args: Delayed<CalcValue<O>>[]) => Delayed<CalcValue<O>>;
 }
 
-function read<O>(trace: Trace, host: O, context: Delayed<CalcValue<O>>, prop: string): Delayed<CalcValue<O>> {
-    return isDelayed(context) ? delay : typeof context === "object" ? trace(context.read(prop, host)) : readOnNonObjectError;
-}
+class CoreRuntime {
```

Comment:
Only reason for using a class here is reduction memory footprint of creating closures for each call of a formula

Reply:
I like it.